### DISCUSSION TITLE
what is the difference between hf unilm and microsoft unilm

### DISCUSSION THREAD

**Discussion by huangrenjie0** [microsoft/fastseq#97] [General]
does the optimization also apply to https://github.com/microsoft/unilm? do you also implement optimization similar to faster transformer?

**Comment by JiushengChen** [comment]
@fuliucansheng, could you help answer this question?

**Reply by fuliucansheng** [reply]
Sure, so sorry for the late reply. I have missed the email.

**Comment by fuliucansheng** [comment]
1. The official unilm repo has more tasks like unilm-v1/unilm-v2/infoxlm pretrain & finetune on different benchmarks and more options for model training experiments. The hf unilm in the fastseq repo is mainly used for faster generation inference. Of course, the inference results are the same.
2. Fastseq is quite different from the repo NVIDIA/FasterTransformer. FasterTransformers is a C++/CUDA repo, and it has GPU kernels/memory optimization and something like that. Fastseq is a Python repo that is easy to use, and the optimization including cache/beam search/io/attention is mainly by python with C/CUDA extensions.
